[twhjames] iP

twhjames opened the pull request:
# Helix
Helix is an intuitive chatbot that helps you efficiently manage tasks, set deadlines, and plan events.
> "We are what we repeatedly do. Excellence, then, is not an act, but a habit." - [Will Durant](https://dailystoic.com/we-are-what-we-repeatedly-do/)

## Why Helix?
- Simple and Intuitive
- ~Fast~ _SUPER_ Fast
- Completely Free

## Download
1. Ensure you have **JDK 17 Installed.**
2. Download Helix Jar file from [here](https://github.com/twhjames/ip/releases/tag/A-Jar).
3. Double-click to open.
4. Add your tasks.
5. Let Helix handle the rest for you 😎.

## Features
- [x] Add tasks, i.e. To-Dos, Deadlines, Events
- [x] Mark tasks as done or undo
- [x] Find tasks by description
- [x] Delete tasks
- [x] View all tasks
- [x] Save tasks to a text file.
- [x] Exit with a farewell message
- [ ] Reminders (coming soon)
- [ ] GUI (coming soon)

## Command Usage
| Command       | Description                                                                 |
|---------------|-----------------------------------------------------------------------------|
| `todo <description>` | Adds a simple to-do helix.task. Example: `todo Read a book`                   |
| `deadline <description> /by <due date>` | Adds a helix.task with a deadline. Example: `deadline Submit report /by 2025-01-30` |
| `event <description> /from <start time> /to <end time>` | Adds an event helix.task. Example: `event Team meeting /from 10am /to 12pm` |
| `list`        | Displays all the tasks in your list.                                        |
| `mark <helix.task number>` | Marks the specified helix.task as done. Example: `mark 1`                      |
| `unmark <helix.task number>` | Marks the specified helix.task as not done. Example: `unmark 1`              |
| `delete <helix.task number>` | Deletes the specified helix.task. Example: `delete 1`                       |
| `bye`         | Exits the chatbot.                                                         |


## Example Interaction
```plaintext
╔════════════════════════════════════════════════════════╗
║                                                        ║
║             _   _  _____  _      ___ __  __            ║
║            | | | || ____|| |    |_ _|\ \/ /            ║
║            | |_| ||  _|  | |     | |  \  /             ║
║            |  _  || |___ | |___  | |  /  \             ║
║            |_| |_||_____||_____||___|/_/\_\            ║
║                                                        ║
╚════════════════════════════════════════════════════════╝

🤖 [helix.Helix] : Hello! I'm your personal assistant, helix.Helix.
🤖 [helix.Helix] : What can I do for you today?

👤 [You]   : todo Read a book
════════════════════════════════════
🗂️  Task Added!
════════════════════════════════════
  📋 Type: TODO
  📝 Description: Read a book

You now have 1 helix.task(s) in your list.
════════════════════════════════════
```

## Code Referece
If you’re exploring the codebase, start from the `main` method in `Helix.java`:
```java
/**
 * The main entry point of the Helix application.
 *
 * @param args command-line arguments (not used in this application)
 */
public static void main(String[] args) {
    new Helix("data/helix_tasklist.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

twhjames committed:
Implement Task 0: Rename, Greet, Exit

- Renamed chatbot to "Helix" and updated all references in the code
- Ensured compliance with Task 0 requirements:
  - Added a welcome message with ASCII art to greet the user
  - Displayed a farewell message on exiting

twhjames committed:
feat: Implement Level 1 Echo functionality

- Enabled chatbot to echo user commands
- Added graceful exit on “bye”

twhjames committed:
feat: Add and List commands for task management

- Implemented the 'add' command to allow users to add tasks to a task list
- Implemented the 'list' command to display all tasks in the task list
- Refactored code to follow Object-Oriented Programming (OOP) principles

twhjames committed:
feat: Add ability to mark tasks as done or not done

- Implemented the functionality to mark tasks as done
- Added the ability to unmark tasks, changing their status to not done
- Updated the task list display to reflect task completion statuses
- Used the Task class to represent tasks, aligning with the recommended A-Classes extension

twhjames committed:
feat: Implement task types for ToDos, Events, and Deadlines

- Introduced `Todo`, `Deadline`, and `Event` classes extending the `Task` superclass
- Added functionality for managing tasks with different attributes:
  - `Todo`: Tasks without a date/time
  - `Deadline`: Tasks with a specific due date
  - `Event`: Tasks with a start and end time
- Enabled polymorphism to store and handle all tasks in a unified data structure
- Updated `toString()` and `getTaskDetails()` for each task type to provide formatted outputs

twhjames committed:
test: add and standardise input/output redirection tests for chatbot

- Updated `input.txt` to simulate user commands for testing the chatbot
- Updated `actual.txt` to define the expected output for the given inputs

twhjames committed:
feat: handle errors for invalid input in chatbot commands

- Added error handling for various user input errors:
  - `TooManyArgumentsException`: Handles extraneous arguments in commands
  - `MissingArgumentException`: Handles missing arguments for commands
  - `InvalidNumberFormatException`: Handles invalid number formats for commands like `mark` and `unmark`
  - `InvalidCommandException`: Handles unknown or unsupported commands
  - `TaskIndexOutOfBoundsException`: Handles invalid task index references
- Improved error messages to provide user-friendly and actionable feedback
- Updated JavaDocs to document exceptions and their usage in the relevant methods
- Ensured consistent input validation across all chatbot commands

twhjames committed:
feat: add support for deleting tasks from the task list

- Implemented the `DeleteCommand` to handle task deletion by index
- Added input validation for the `delete` command, including:
  - `MissingArgumentException` for missing task numbers
  - `TooManyArgumentsException` for extraneous arguments
  - `InvalidNumberFormatException` for non-numeric task numbers
  - `TaskIndexOutOfBoundsException` for invalid task indices
- Updated `TaskList` to support task removal with the `removeTask` method
- Updated `CommandFactory` to create `DeleteCommand` with proper validation

twhjames committed:
feat: Implement Enums for Command Type, Task Type, Task Status, Output Symbols, and Execution Status

- Added enum classes: `CommandType`, `TaskType`, `TaskStatus`, `OutputSymbol`, and `ExecutionStatus`
- Updated all relevant classes and methods to utilise the new enums for improved clarity, consistency, and maintainability
- Refactored command creation logic in `CommandFactory` to integrate `CommandType` enums
- Replaced hardcoded symbols and task statuses with enum references in command implementations

twhjames committed:
feat(delete): update delete command output and TextUiTesting

- Refined the output message for the delete command to improve clarity
- Updated TextUiTesting to cover changes in the delete command output

twhjames committed:
docs: update README with detailed guide and examples

- Added an overview section describing Helix's purpose and functionality
- Included detailed setup instructions under "Getting Started"
- Expanded "Usage" section with comprehensive command examples in tabular format
- Introduced a "Project Structure" section to showcase the code organisation
- Added an example interaction session to illustrate Helix's capabilities

twhjames committed:
Refactor(task): reorganise TaskList and improve code clarity

- Moved TaskList to the Task package for better logical grouping
- Added JavaDocs for enums in the Enums package to enhance documentation
- Updated comments for improved readability
- Removed unnecessary imports to clean up the codebase

twhjames committed:
Docs(enums): add JavaDocs for enums and clean up code

- Added JavaDocs to CommandType, ExecutionStatus, OutputSymbol, TaskStatus, and TaskType
- Improved documentation for enum constants, fields, and methods

twhjames committed:
Feat(storage): implement persistent task saving and loading

- Added functionality to save tasks to a file automatically whenever the task list changes.
- Tasks are now loaded from a specified file (e.g., `./data/helix_tasklist.txt`) on startup.
- Ensured compatibility with relative paths for cross-platform support.
- Implemented fallback to create the file and necessary directories if they do not exist.
- Handled the scenario where the data file is missing, creating a fresh file in such cases.
- Added error handling for corrupted or malformed data file content.

twhjames committed:
Merge branch 'branch-Level-7' into master

twhjames committed:
feat(Level-8): Add support for parsing and storing tasks with detailed dateTime formats

- Updated `Storage` to parse events and deadlines using the dateTime format `MMM dd yyyy, h:mm a`.
- Extended `Deadline` and `Event` classes to support the input formats `d/M/yyyy HHmm` and `yyyy-MM-dd HHmm`, with output in `MMM dd yyyy, h:mm a`.
- Improved event parsing in `Storage.parseTask` to correctly handle event start and end dateTime values.
- Added validation to ensure that start dateTime is before end dateTime for events.
- Updated serialisation logic to store deadlines and events using the new dateTime format.

twhjames committed:
Merge branch 'branch-Level-8 into master

twhjames committed:
feat: Enhance file path management and improve OOP structure

- Introduced `FilePath` enum to centralise and manage file paths.
- Refactored `Helix` constructor to dynamically resolve file paths based on the runtime environment (project root vs `text-ui-test`).
- Made the code more OOP-compliant by:
  - Abstracting file path logic to `Helix`.
  - Delegating responsibilities to classes like `Storage` and `TaskList`.
  - Using polymorphism for command execution in `Command` and its subclasses.
- Improved code readability, maintainability, and extensibility.

This update ensures better adherence to OOP principles while simplifying file path management for dynamic use cases.

twhjames committed:
Merge branch 'A-MoreOOP'

twhjames committed:
refactor: Restructure project packages and improve file path handling

- Organised classes into packages under `helix` for better modularity:
  - Added `helix.command` for command-related classes.
  - Added `helix.enums` for enums.
  - Added `helix.exception` for custom exceptions.
  - Added `helix.storage` for storage management.
  - Added `helix.task` for task-related classes.
  - Added `helix.ui` for UI-related classes.
- Ensured compatibility with `runtest.sh` and improved maintainability.

twhjames committed:
Merge branch 'branch-A-Packages' into master

twhjames committed:
Merge branch 'add-gradle-support' into master

twhjames committed:
feat: Add Gradle support to project

twhjames committed:
test: Add JUnit tests for core components and storage

- Added tests for `TaskList::addTask_validTask_taskAdded` and `TaskList::removeTask_validIndex_taskRemoved`
  to verify task addition, removal, and interaction with storage.
- Added tests for `Storage::save_tasksWithSingleTask_taskPersistedSuccessfully` and
  `Storage::saveAndLoad_emptyTaskList_emptyListRetrieved` to validate task persistence and retrieval functionality.
- Introduced `StubStorage` to eliminate dependencies on file-based storage during unit tests.

twhjames committed:
Merge branch 'branch-A-JUnit' into master

twhjames committed:
docs: Add JavaDocs for TaskListTest, StubStorage, and StorageTest

- Added detailed JavaDoc comments for `TaskListTest` methods to describe their purpose and behaviour.
- Documented `StubStorage` class to explain its role as a mock implementation for isolating tests from storage dependencies.
- Added comprehensive JavaDocs for `StorageTest` to outline the purpose and structure of each test method.
- Improved overall code documentation for better readability and maintainability.

twhjames committed:
Merge branch 'branch-A-JavaDoc' into master

twhjames committed:
chore(imports): Reformat import order to follow coding standards

- Aligned import statements with project coding guidelines for consistency.
- No functional changes introduced.

twhjames committed:
Merge branch 'branch-A-CodingStandard' into master

twhjames committed:
feat(find): Add `FIND` command support in `CommandType` enum

- Updated `CommandType` enum to include a new `FIND` command.
- Added support for finding tasks based on keywords in their descriptions.

twhjames committed:
Merge branch 'branch-Level-9' into master

twhjames committed:
feat: Integrate Checkstyle and improve project structure

- Added Checkstyle plugin to Gradle for code quality enforcement
- Updated codebase to align with Checkstyle rules and standards
- Restructured test directory and ensured JUnit tests run successfully

delonlee01 commented on src/main/java/helix/Helix.java:
```diff
@@ -0,0 +1,101 @@
+        } else {
+            // running from project root
+            resolvedFilePath = Paths.get(storageFile).toAbsolutePath().normalize().toString();
+        }
```
Would it better to separate the storage file, instead of sharing?

delonlee01 commented on src/main/java/helix/task/Deadline.java:
```diff
@@ -0,0 +1,95 @@
+     * @return A LocalDateTime object representing the parsed due date.
+     * @throws IllegalArgumentException if the input does not match any expected format.
+     */
+    private LocalDateTime parseDueDate(String dueDate) {
```
```suggestion
    private LocalDateTime parseDueDate(String dueDate) throws IllegalArgumentException {
```

delonlee01 commented on src/main/java/helix/task/Event.java:
```diff
@@ -0,0 +1,101 @@
+     * @return a LocalDateTime object representing the parsed date-time
+     * @throws IllegalArgumentException if the input string does not match any expected format
+     */
+    private LocalDateTime parseDateTime(String dateTime) {
```
```suggestion
    private LocalDateTime parseDateTime(String dateTime) throws IllegalArgumentException {
```

delonlee01 commented on src/main/java/helix/Helix.java:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @param filePath the file path to the storage file where tasks are saved and loaded
+     */
+    public Helix(String filePath) {
```
Is `filePath` needed? Looks like its not used

delonlee01 reviewed (commented):
Accident. Please ignore.

twhjames committed:
chore(checkstyle): Add Checkstyle for code quality enforcement

- Integrate Checkstyle to enforce coding standards
- Add Checkstyle configuration file
- Update build scripts to include Checkstyle checks
- Ensure compliance with initial Checkstyle rules

twhjames committed:
Merge branch 'branch-A-Checkstyle' into master

RT0118 reviewed (commented):
Great work on Helix! It's good that all variables use the correct case (camelcase, pascalcase, etc). The switch cases indentation also follow the proper indentation. Plus, it follows the use of proper indentation as recommended. And with the limiting of line length and use of egyptian braces, makes the code super readable! Lastly, it's amazing that you took the time to do javadocs for all private and public functions.

amir-dha commented on src/main/java/helix/exception/TooManyArgumentsException.java:
```diff
@@ -0,0 +1,25 @@
+     * @param command the helix.command name
+     * @param exampleFormat an example of the correct helix.command format
+     */
+    public TooManyArgumentsException(String command, String exampleFormat) {
```
Your error names are really specific. This is great for understanding the code!

amir-dha commented on src/main/java/helix/task/Todo.java:
```diff
@@ -0,0 +1,49 @@
+     * @return An empty string as Todo tasks have no additional details.
+     */
+    @Override
+    public String getTaskDetails() {
```
Perhaps, a shorter name will do?

amir-dha commented on src/main/java/helix/ui/Ui.java:
```diff
@@ -0,0 +1,267 @@
+ */
+public class Ui {
+
+    private final String logo = """
```
perhaps, constants like this could be written in capital letters

amir-dha reviewed (commented):
Your code generally does a great job in following code quality standards! It was a great read, thanks!

twhjames committed:
feat(gui): Add JavaFX-based GUI for Helix chatbot

- Integrated a GUI for the Helix chatbot using JavaFX.
- Implemented DialogBox and MainWindow for handling user and Helix interactions.
- Added Launcher class to handle application startup as per JavaFX guidelines.
- Configured MainApp to initialise and display the main application window.
- Included FXML and CSS files for defining and styling the GUI components.

twhjames committed:
Merge branch 'branch-Level-10' into master

twhjames committed:
style(gui): Improve UI styling

- Enhanced the graphical interface for better user experience.
- Fixed minor UI inconsistencies.
- Improved layout and visual hierarchy.

twhjames committed:
Merge 'branch-Level-10' into master

twhjames committed:
Build: Update build.gradle for JAR packaging

- Ensure the correct Main-Class is set in the manifest

twhjames committed:
Chore: Add assertions to ensure internal consistency

- Introduce assertions to catch unexpected null values and enforce assumptions within critical components.
- Enhance code reliability by preventing potential NullPointerException and ensuring expected conditions hold true during execution.

twhjames committed:
Merge pull request #1 from twhjames/branch-A-Assertions

Chore: Add assertions to ensure internal consistency

twhjames committed:
Merge master into 'branch-A-CodeQuality'

twhjames committed:
Refactor: Abstract validation logic and standardise constants

- Extract task number validation logic for `mark`, `unmark`, and `delete` commands in `CommandFactory` to reduce duplication.
- Standardise constant names in `ConsoleUi` to follow uppercase naming conventions.

twhjames committed:
Merge pull request #2 from twhjames/branch-A-CodeQuality

Refactor: Improve code quality and enforce best practices

twhjames committed:
Build: Add gradle.yml for CI pipeline

- Add a GitHub Actions workflow to automate Gradle builds.

twhjames committed:
Style: Rename constants to camelCase to comply with CheckStyle rules

- Modify constant names from UPPER_CASE to camelCase as checkstyleMain flagged them.

twhjames committed:
Fix(ci): Install dos2unix on Ubuntu to prevent runtest.sh error

- The Ubuntu runner does not have dos2unix pre-installed, causing runtest.sh to fail.
- This update installs dos2unix before executing the script.

twhjames committed:
Chore(ci): Remove dos2unix installation and IO redirection tests

- The current program does not rely on IO redirection tests, making dos2unix unnecessary.
- This commit removes the redundant steps from the GitHub Actions workflow.

twhjames committed:
Merge pull request #3 from twhjames/branch-A-CI

Build: Add gradle.yml for CI pipeline

twhjames committed:
Feat(update-task): Add update functionality for tasks

- Implement `UpdateCommand` to handle task updates.
- Validate task index and ensured task type consistency before updating.
- Update `TaskList` to include `updateTask` method for modifying tasks.
- Refine `Deadline` and `Event` classes to properly parse and update details.

twhjames committed:
Merge pull request #4 from twhjames/branch-C-Update

Feat(update-task): Add update functionality for tasks

twhjames committed:
Docs: Update Javadoc comments to follow project standards

Refactor Javadoc comments for clarity and consistency.

twhjames committed:
Refactor(core): Enforce SLAP in codebase

- Ensure methods operates at a single level of abstraction.
- Ensure code maintainability by reducing redundant logic.

twhjames committed:
Feat(ui): Improve GUI for better error visibility and user experience

Ensure consistent theme and enhance UI responsiveness.

twhjames committed:
Merge branch 'branch-A-BetterGui' into master

twhjames committed:
Feat(ui): Refactor MainWindow to follow SLAP and improve UX

- Extracted helper methods for better separation of concerns.
- Ensured SLAP compliance by making each method handle a single task.
- Improved UX by clearing `userInput` after command execution.

twhjames committed:
Merge branch 'branch-A-Personality' into master

twhjames committed:
Fix(exception): Improve InvalidCommandException message clarity

- Lists valid commands in the error message for better guidance.
- Helps users quickly identify correct commands and avoid confusion.

twhjames committed:
Docs: Update README with detailed User Guide

- Refactor structure with installation, usage, and command details.
- Improve readability and consistency.

twhjames committed:
Merge branch 'branch-A-UserGuide' into master

twhjames committed:
Docs: Remove contributing section from README user guide

Remove the contributing section as it is irrelevant to the user guide.

twhjames committed:
Merge master into branch 'branch-A-UserGuide'

twhjames committed:
Docs: Tidy up README user guide for clarity and consistency

- Improve consistency in wording and style.

twhjames committed:
Merge branch 'branch-A-UserGuide' into master

twhjames committed:
Feat(docs): Improve README for project website clarity

- Refine explanations and formatting for better readability.
- Update installation and usage instructions for clarity.
- Refine file storage section with structured format and examples.

twhjames committed:
Merge pull request #6 from twhjames/branch-A-UserGuide

Feat(docs): Improve README for project website clarity
